File: demos/PytorchComparison/torch-basic-nn.py
# Do the same as torch-basic-nn.cpp but in Python using PyTorch
# To run: python torch-basic-nn.py

import torch
import torch.nn as nn
import torch.optim as optim
import numpy as np

M = 100 # Number of samples
N = 100   # Number of features
N_ITER = 5000 # Number of iterations

# Create a simple neural network with 3 hidden nodes and 1 output node
class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.fc1 = nn.Linear(N, 3)
        self.fc2 = nn.Linear(3, 1)
    def forward(self, x):
        x = torch.relu(self.fc1(x))
        x = self.fc2(x)
        return x

def main():
    # Create a simple dataset
    x = np.random.randn(M, N).astype(np.float32)
    y = np.random.randn(M, 1).astype(np.float32)

    # Convert the dataset to PyTorch tensors
    x = torch.from_numpy(x)
    y = torch.from_numpy(y)

    # Create the neural network
    net = Net()

    # Create the loss function and the optimizer
    criterion = nn.MSELoss()
    optimizer = optim.SGD(net.parameters(), lr=0.01)

    # Print the initial loss
    output = net(x)
    loss = criterion(output, y)
    print("Initial loss: ", loss.item())

    # Calculate start time
    import time
    start = time.time()

    # Train the neural network
    for i in range(N_ITER):
        optimizer.zero_grad()
        output = net(x)
        loss = criterion(output, y)
        loss.backward()
        optimizer.step()

    # Stop the timer
    end = time.time()

    # Print the final loss
    output = net(x)
    loss = criterion(output, y)
    print("Final loss: ", loss.item())
    print("Time: ", end - start, " seconds")
    

if __name__ == "__main__":
    main()

File: demos/PytorchComparison/torch-basic-nn.cpp
// Do the same thing what's done in basic-nn.cpp, but using libtorch instead of
// Clad.

/*
  To build and run
  1. Create a build directory: mkdir build
  2. Change to the build directory: cd build
  3. Run cmake: cmake .. -DCMAKE_PREFIX_PATH=/path/to/libtorch
  4. Run make: make
  5. Run the executable: ./torch-basic-nn
*/

#include <chrono>
#include <iostream>
#include <random>
#include <torch/torch.h>

#define M 100       // num of data points
#define N 100       // num of features
#define N_ITER 5000 // num of iterations

// C++ neural network model defined with value semantics
struct TwoLayerNet : torch::nn::Module {
  // constructor with submodules registered in the initializer list
  TwoLayerNet(int64_t D_in, int64_t D_out, int64_t H)
      : linear1(register_module("linear1", torch::nn::Linear(D_in, H))),
        linear2(register_module("linear2", torch::nn::Linear(H, D_out))) {}

  torch::Tensor forward(torch::Tensor x) {
    x = torch::relu(linear1->forward(x));
    x = linear2->forward(x);
    return x;
  }
  torch::nn::Linear linear1;
  torch::nn::Linear linear2;
};

int main() {
  // Generate random data
  // std::default_random_engine generator;
  // std::normal_distribution<double> distribution(0.0, 1.0);
  torch::Tensor x = torch::randn({M, N});
  torch::Tensor y = torch::randn({M, 1});

  // Benchmark the time for 100 matrix multiplications
  torch::Tensor z;
  auto start_bench = std::chrono::high_resolution_clock::now();
  for (int i = 0; i < 100; ++i)
    z = torch::matmul(x, x);
  auto end_bench = std::chrono::high_resolution_clock::now();
  std::chrono::duration<double> elapsed_bench = end_bench - start_bench;
  std::cout << "Time taken for 100 matrix multiplications: "
            << elapsed_bench.count() << " seconds" << std::endl;

  // Initialize model weights
  TwoLayerNet model(N, 1, 3);

  // Create the optimizer and loss function
  torch::optim::SGD optimizer(model.parameters(), 0.01);
  torch::nn::MSELoss loss_fn;

  // Calculate the loss before optimization
  torch::Tensor output = model.forward(x);
  std::cout << "Initial loss before optimization: "
            << loss_fn(output, y).item<double>() << std::endl;

  // Optimize
  auto start = std::chrono::high_resolution_clock::now();
  for (int i = 0; i < N_ITER; ++i) {
    optimizer.zero_grad();
    torch::Tensor output = model.forward(x);
    torch::Tensor loss = loss_fn(output, y);
    loss.backward();
    optimizer.step();
  }
  auto end = std::chrono::high_resolution_clock::now();
  std::chrono::duration<double> elapsed = end - start;

  // Calculate the loss after optimization
  output = model.forward(x);
  std::cout << "Final loss after optimization: "
            << loss_fn(output, y).item<double>() << std::endl;
  std::cout << "Time taken: " << elapsed.count() << " seconds" << std::endl;
  return 0;
}
